### PR TITLE
[ML] Wait for controller to respond to commands

### LINKED ISSUE #62823: [ML] Add a response mechanism to ML controller command processing
When the ML Java code needs to start one of the ML native processes (`autodetect`, `normalize` or `data_frame_analyzer`) it sends a command to the `controller` process telling it to spawn the required process.  Currently the communications are one way only - the JVM sends a command to the controller and assumes it will be actioned immediately.  There is no mechanism for the controller to respond when it has actioned the command.  This seemed reasonable in the initial design because the controller is completely dedicated to starting and killing processes, and these were assumed to be very fast operations.

We have observed that when security software is running on a machine spawning a new process can take a very long time - over 20 seconds has been observed between the command being received in the controller and the resulting `posix_spawn` call returning.  This invalidates the assumption that commands issued to `controller` by the JVM will be near instantaneous.  It causes a problem because the timeout waiting for the named pipes to connect starts immediately after the command is issued, but the process may not actually start until considerably later.

Therefore, there is a need for `controller` to be able to report back to the ES JVM when each command sent to it has been actioned.  Then the ES JVM should not try to connect the named pipes to a process until the controller has reported that it has actually spawned that process.  This will mean that the configured timeout for connecting the named pipes is measured from a more appropriate point in time.

### PR DESCRIPTION
**Pull request by droberts195**
This change makes threads that send a command to the ML
controller process wait for it to respond to the command.
Previously such threads would block until the command was
sent, but not until it was actioned.  This was on the
assumption that the sort of commands being sent would be
actioned almost instantaneously, but that assumption has
been shown to be false when anti-malware software is
running.

Relates elastic/ml-cpp#1520
Fixes #62823

### CODE REVIEW

**Comment by droberts195** [comment]
I have marked this `>non-issue` as it will be jointly release-noted with elastic/ml-cpp#1520.

The hardest thing about this change is merging it without completely breaking CI while the Java and C++ builds are out-of-sync.  The basic plan is:

1. Open another PR that will mute every single ML test that uses native code
2. Merge that muting PR late on a Friday evening
3. Retest the C++ PR and check that it passes with the Java tests muted
4. Merge the C++ PR on the Saturday
5. Wait for the snapshot build of the C++ to complete
6. Merge master into this PR and revert the commit that muted all the ML native tests
7. Merge this PR after it goes green, thus reenabling the ML native tests

The intention is that step 7 is complete by the end of Sunday so that ML test coverage is back to normal by Monday morning.

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/NativeController.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/NativeController.java:224]: I am curious on locking on this and not on `ourResponseTracker` for the entire while loop.

If we lock on this, that means that any other command could not be processed. 

Our commands are always handled in a single thread and could NEVER be processed out of order correct? If they are ever processed out of order, this seems like it would cause a thread lock.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/NativeController.java
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/NativeController.java:192]:
```suggestion
            if (awaitCompletion) {
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/NativeAutodetectProcessFactory.java
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/NativeAutodetectProcessFactory.java:132]: I was matching the format of the existing message from a few lines below.  But I guess I can change the pre-existing ones too.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/NativeController.java
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/NativeController.java:222]: It's correct that at the moment the C++ always processes commands in order.  This may not always be the case though.

It's also true that in the future if two commands were processed out-of-order in the C++ then this Java code would delay one of the threads waiting for a response unnecessarily.  However, it wouldn't be a deadlock - it would just mean threads 1 and 2 would unblock on the response to command 1 (or disconnection of `controller`) even though thread 2 could have unblocked on the earlier response to command 2.

So some optimisation work could be done here if the C++ is ever modified to process commands out-of-order.

The actual parsing needs to synchronize on `responseIteratorHolder`, because only one thread can be parsing the stream at any time.  What _could_ be done is that the other threads simply wait for their response latches on the assumption that the thread doing the parsing will count these down.  _However_, this is non-trivial, because if the thread doing the parsing has just decided that it can stop parsing but hasn't exited the synchronized block then _no_ thread would be parsing.  So it would need more complex synchronization to ensure that the decision of one thread to wait for its latch instead of for the parser lock would need to be atomic with the thread currently holding the parser lock to decide to stop parsing (which is _not_ the same as releasing the parser lock).  Rather than risk messing this up I went for the simpler approach.

The alternative is to have a dedicated thread for parsing responses like we do for most C++ outputs, but this seems wasteful in the case of `controller` where there is very low traffic and we'd be permanently hogging a thread for it.

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/NativeController.java:222]: > because only one thread can be parsing the stream at any time.

👍 

Thanks for the additional information. It just wasn't obvious to me :D

**Review by benwtrent — APPROVED** [review]

**Comment by droberts195** [comment]
Jenkins test this

**Comment by droberts195** [comment]
Jenkins test this again

**Pull request merged**
